tests/test_qdatabase_pgsql.c

```diff
@@ -0,0 +1,334 @@
+*/
+void test_create_table_animals(qdb_t* database) {
+    int rows;
+
+    rows = database->execute_update(database,
+        "CREATE TABLE animals ("
+        "animal_id INT PRIMARY KEY,"
+        "animal_name VARCHAR(255),"
+        "species VARCHAR(255),"
+        "gender VARCHAR(255),"
+        "age INT,"
+        "weight INT,"
+        "description VARCHAR(255)"
+        ");"
+    );
+    ASSERT_EQUAL_INT(rows, 0);
+
+    rows = database->execute_updatef(database,
+        "INSERT INTO animals VALUES"
+        "(%s, %s, %s, %s, %s, %s, %s),"
+        "(%s, %s, %s, %s, %s, %s, %s),"
+        "(%s, %s, %s, %s, %s, %s, %s),"
+        "(%s, %s, %s, %s, %s, %s, %s),"
+        "(%s, %s, %s, %s, %s, %s, %s);",
+            "1", "'Dog'",     "'Canis lupus familiaris'",   "'Male'",   "2", "15",  "'A friendly dog named Max'",
+            "2", "'Cat'",     "'Felis catus'",              "'Female'", "5", "8",   "'A lazy cat named Lucy'",
+            "3", "'Rabbit'",  "'Oryctolagus cuniculus'",    "'Male'",   "1", "2",   "'A cute little rabbit named Bugs'",
+            "4", "'Hamster'", "'Mesocricetus auratus'",     "'Female'", "2", "50",  "'A hyperactive hamster named Sparky'",
+            "5", "'Fish'",    "'Poecilia reticulata'",      "'Female'", "1", "0.1", "'A peaceful betta fish named Blue'");
+    ASSERT_EQUAL_INT(rows, 5);
+}
+
+TEST("Test1: qdb()/open()/close()/free()") {
+    qdb_t *db = NULL;
+
+    /* test qdb() */
+    db = qdb(
+        default_dbtype,
+        default_addr,
+        default_port,
+        default_database,
+        default_username,
+        default_password,
+        default_autocommit
+    );
+
+    ASSERT_NOT_NULL(db);
+    ASSERT_FALSE(db->connected);
+    ASSERT_TRUE(db->open(db));
+    ASSERT_TRUE(db->connected);
+    ASSERT_TRUE(db->close(db));
+    ASSERT_FALSE(db->connected);
+    db->free(db);
+}
+
+TEST("Test2: execute_updatef()") {
+    qdb_t *default_db = test_connect_database(default_database);
+    ASSERT_TRUE(default_db->connected);
+    qdb_t *test_db = test_connect_testdb(default_db);
+
+    /* test: execute_updatef */
+    test_create_table_animals(test_db);
+
+    /* disconnect */
+    ASSERT_TRUE(test_db->close(test_db));
+    ASSERT_TRUE(default_db->close(default_db));
+
+    /* free */
+    test_db->free(test_db);
+    default_db->free(default_db);
+}
+
+TEST("Test3: execute_queryf()") {
+    qdb_t *default_db = test_connect_database(default_database);
+    ASSERT_TRUE(default_db->connected);
+    qdb_t *test_db = test_connect_testdb(default_db);
+    test_create_table_animals(test_db);
+
+    /* test: execute_queryf  */
+    qdbresult_t *qrst = test_db->execute_queryf(test_db, "SELECT * FROM %s;", "animals");
+    ASSERT_EQUAL_INT(qrst->get_rows(qrst), 5);
+    ASSERT_EQUAL_INT(qrst->get_cols(qrst), 7);
+    ASSERT_EQUAL_INT(qrst->get_row(qrst),  0);
+    ASSERT_EQUAL_STR(qrst->get_str(qrst, "animal_name"), "Dog");
```

Comment:
execute_query() shouldn't fetch any row yet before get_next() is called.
get_next() needs to be executed before this line.